### PR TITLE
Refactored redirects handling in auth views.

### PR DESCRIPTION
**Pull request by aaugustin**
There's a lot of security-sensitive code around handling redirects in LoginView and LogoutView, in django.contrib.auth.views.

The actual behavior is very similar in both classes but the implementation is quite different, making it hard to follow what's going on.

I refactored to align the implementation in both classes and move the common parts to a mixin (which existed already).

This PR builds upon #15607. It is split in atomic commits to show the refactoring steps. I recommend you review it commit-by-commit.

### CODE REVIEW

**Comment by aaugustin** [comment]
I rebased on main after #15607 was merged.

**Comment by felixxm** [comment]
@aaugustin Thanks :+1: I merged the 2nd cleanup in 903702dfb1461cbb4ab93f841651e6882a72d337.

> Rename SuccessURLAllowedHostsMixin to RedirectURLMixin.
> This is a private API. However, there are references to this API
> on StackOverflow so we may consider providing an alias

It's a private API but folks probably use it. I'd add a note to the ["Backwards incompatible changes in 4.1 -> Miscellaneous"](https://docs.djangoproject.com/en/dev/releases/4.1/#miscellaneous) section, e.g.:
```
* The undocumented ``django.contrib.auth.views.SuccessURLAllowedHostsMixin``
  mixin is replaced by ``RedirectURLMixin``.
```
